src/grammar.adoc

```diff
@@ -0,0 +1,208 @@
+_module-body_ `)`
+|catalog-key         |::=| `(` `*catalog_key*` _catalog-name_ _catalog-version_ `)`
+|===
+
+
+==== Module Bodies
+
+[{bnf}]
+|===
+|module-body |::=|  _dependency_* _symtab_? _macro-alias_* _module-mactab_?
+
+|macro-alias |::=|  `(` `*alias*` _macro-name_ _macro-ref_ `)`
+
+|module-mactab   |::=|  `(` `*macro_table*` _macro-or-export_* `)`
+|macro-or-export |::=|  _macro-defn_  \|  _export_
+
+|export      |::=|  `(` `*export*` _export-item_* `)`
+|export-item |::=|  _macro-ref_                                         +
+                \|  `(` `*from*` _module-name_ _name-or-address_* `)`   +
+                \|  `(` `*all_from*` _module-name_ `)`
+|===
+
+
+=== Macro Definitions
+
+[{bnf}]
+|===
+|macro-defn |::=|  `(` `*macro*` _macro-name_? _signature_ _template_ `)`
+
+|signature   |::=|  _param-specs_ _result-spec_?
+|param-specs |::=|  `(` _param-spec_* _rest-spec_? `)`  \|  `[` _param-spec_* _rest-spec_? `]`
+|param-spec  |::=|  _param-name_  \|  _simple-param-spec_  \|  _grouped-param-spec_
+|param-name  |::=|  _unannotated-identifier-symbol_
+
+|simple-param-spec  |::=|  `(` _param-name_  _base-shape_ _simple-cardinality_ `)`
+|base-shape         |::=| _tagged-type_  \|  _primitive-type_  \|  _macro-ref_
+|simple-cardinality |::=|  `*!*`  \|  `*+*`  \|  `*'?'*`  \|  `*'{asterisk}'*`
+
+|grouped-param-spec  |::=|  `(` _param-name_ `[` _base-shape_ `]` _grouped-cardinality_? `)`
+|grouped-cardinality |::=|  `*+*`  \|  `*'{asterisk}'*`
+
+|rest-spec        |::=|  `(` _param-name_ _base-shape_ _rest-cardinality_  `)`
+|rest-cardinality |::=|   `*\...*`  \|  `*\...+*`
+
+|tagged-type   |::=| _abstract-type_ \| _concrete-type_
+|abstract-type |::=| `*any*` \| `*number*` \| `*exact*` \| `*text*` \| `*lob*` \| `*sequence*`
+|concrete-type |::=|  `*'null'*`  \|  `*bool*`  \|  `*timestamp*`  \|  `*int*`  \|  `*decimal*`
+                  \|  `*float*`  \|  `*string*`  \|  `*symbol*`  \|  `*blob*`  \|  `*clob*`
+                  \|  `*list*`  \|  `*sexp*`  \|  `*struct*`
+
+|primitive-type |::=|  `*var_symbol*`  \|  `*var_string*`
+                   \|  `*var_int*`  \|  `*var_uint*`
+                   \|  `*uint8*`    \|  `*uint16*`   \|  `*uint32*`  \|  `*uint64*`
+                   \|  `*int8*`     \|  `*int16*`    \|  `*int32*`   \|  `*int64*`
+                   \|  `*float16*`  \|  `*float32*`  \|  `*float64*`
+
+|result-spec |::=|  `*\->*` _tagged-type_ _simple-cardinality_
+|===
+
+
+=== Template Expression Language
+
+[{bnf}]
+|===
+|template |::=|  _identifier_  \|  _literal_  \|  _quasi-literal_
+             \|  _special-form_  \|  _macro-invocation_
+
+|literal       |::=|  _null_  \|  _bool_  \|  _int_  \|  _float_  \|  _decimal_  \| _timestamp_
+                  \|  _string_  \|  _blob_  \|  _clob_
+
+|quasi-literal |::=|  `[` _template_* `]`  \|  `{` _quasi-field_* `}`
+|quasi-field   |::=| _text_ `:` _template_
```

Comment:
From a notation point of view, there is something weird about omitting commas for lists and structs, but not the colon here.  I know it's because we're in a grammar over the data model, but still it is odd to me because we are conflating this in the BNF.